src/workerd/api/sockets.h

```diff
@@ -69,27 +72,48 @@ struct SocketOptions {
 struct InitData {
   jsg::Ref<ReadableStream> readable;
   jsg::Ref<WritableStream> writable;
-  IoOwn<kj::PromiseFulfillerPair<void>> closeFulfiller;
+  IoOwn<kj::PromiseFulfillerPair<bool>> closeFulfiller;
+  IoOwn<kj::PromiseFulfillerPair<void>> jsCloseFulfiller;
+  kj::ForkedPromise<void> jsCloseFulfillerFork;
 };
 
 class Socket: public jsg::Object {
 public:
-  Socket(InitData data)
+  Socket(jsg::Lock& js, InitData data)
       : readable(kj::mv(data.readable)), writable(kj::mv(data.writable)),
-        closeFulfiller(kj::mv(data.closeFulfiller)) {};
-  Socket(kj::Promise<kj::Own<kj::AsyncIoStream>> connectionPromise);
+        closeFulfiller(kj::mv(data.closeFulfiller)),
+        jsCloseFulfiller(kj::mv(data.jsCloseFulfiller)),
+        jsCloseFulfillerFork(kj::mv(data.jsCloseFulfillerFork)) {
+    auto& context = IoContext::current();
+    // Attach a callback to close the readable/writable streams when the socket is closed (either
+    // explicitly or implicitly).
+    context.awaitIo(js, kj::mv(closeFulfiller->promise), [this](
+        jsg::Lock& js, bool isImplicit) mutable {
+      if (!isImplicit) {
+        // A `close` call was made on the socket. Forcibly close the readable/writable streams.
```

Comment:
Why not put this code directly inside the `close()` method, rather than waiting on the promise and having this `isImplicit` flag?

Reply:
The reason is I couldn't figure out how to get the `jsg::Lock` inside the `close()`. I can of course get it just fine when the JS is calling the `close` method, but whenever I want to call it from my AsyncIoStream I couldn't figure out a way to get a hold of this lock and any attempts to store it have failed me.

I'm definitely missing something though so would appreciate tips here.

Reply:
> whenever I want to call it from my AsyncIoStream

Aren't those cases always "implicit" anyway? I specifically meant `close()` as called from JavaScript, i.e. the "explicit" case.